### PR TITLE
feat(dashboard): menu improvements, fallback support for Drill to Detail

### PR DESCRIPTION
**Pull request by codyml**
### SUMMARY

This PR does some code cleanup, adds/updates testing, and extends support for Drill to Detail to additional edge cases.

- **Code DRYing for right-click and chart header dropdown Drill to Detail menus.**  Right-click menu and chart header dropdown menu currently relies on some duplicated code between the `SliceHeaderControls` and `Chart` components.  This PR combines/moves all Drill to Detail code to `Chart/DrillDetail`.
- **Tests for menu items.**  Drill to Detail menu items currently lacks unit tests.  This PR adds some.
- **Basic fallback support for right-click Drill to Detail menus for all viz plugins.**  Right-click Drill to Detail support is currently limited to viz plugins that have been updated to handle the `contextmenu` event.  This PR adds a fallback `contextmenu` event handler to the `ChartRenderer` chart wrapper that allows access to Drill to Detail without any filters, just like what's currently possible from the chart header menu.
- **Updated menu hierarchy.**  The right-click menu currently shows all items as a flat list.  This PR updates the right-click menu to have a top-level "Drill to detail" menu item that opens the D2D modal with no filters and a "Drill to detail by" submenu that contains "Drill to detail by XXX" menu items that open the D2D modal with filters.
- **Support for additional states.**: This PR updates the right-click menu and header dropdown menus to behave as follows in the following situations:
  - **Viz plugin has not been updated to handle the `contextmenu` event:** right-click menu has enabled "Drill to detail" and enabled "Drill to detail by" submenu containing just a message noting the lack of support.  Header dropdown menu has enabled "Drill to detail".
  - **Viz plugin has been updated, but the chart doesn't group data by dimension:** right-click menu has disabled "Drill to detail" with an explanatory tooltip and disabled "Drill to detail by" submenu.  Header dropdown menu has disabled "Drill to detail" with an explanatory tooltip.
  - **Viz plugin has been updated and the chart groups data by dimension:** right-click menu has enabled "Drill to detail" and enabled "Drill to detail by" submenu.  If you right-clicked on a valid dimension value, it'll contain "Drill to detail by" options.  If you didn't, it'll contain an explanatory message.  Header dropdown menu has enabled "Drill to detail".

#### Notes for adding/improving Drill to Detail support for viz plugins
- If a plugin hasn't been updated to explicitly support Drill to Detail, header dropdown menu and right-click menu Drill to Detail will work, but the "Drill to detail by" submenu will contain a message saying the chart doesn't have support.
- To update a plugin to support Drill to Detail, handle the `contextmenu` event such that the `onContextMenu` chart hook gets called with filters corresponding to whatever dimension(s) the user right-clicked on.
- To mark a plugin as supported, add the `Behavior.DRILL_TO_DETAIL` behavior to the plugin's `ChartMetadata` instance.  The Drill to Detail menu component checks for this behavior whenever `onContextMenu` is called without filters to determine if it should show the "not yet supported" message or the "you didn't select a dimension" message.  Therefore, if you add the `Behavior.DRILL_TO_DETAIL` behavior, you can let the fallback handler catch right-clicks to parts of the chart that don't correspond to dimension values and the menu will correctly interpret these clicks as clicks not corresponding to a dimension rather than clicks to an unsupported chart.
- Drill to Detail should be disabled in situations when a viz plugin is not grouping data by dimension, as there aren't any aggregations to de-aggregate.  To indicate when this is the case for a viz plugin, add a `noAggregations` function to the viz plugin's `ChartMetadata` instance that receives `formData` as a param and returns `true` if the chart contains only atomic data.

cc: @kasiazjc @michael-s-molina 

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

1. Unsupported chart, header menu
2. Unsupported chart, right-click menu
3. Supported chart w/o aggregations, header menu
4. Supported chart w/o aggregations, right-click menu
5. Supported chart w/ aggregations, header menu
6. Supported chart w/ aggregations, right-click menu on empty space
7. Supported chart w/ aggregations, right-click menu on dimension

**Before**

https://user-images.githubusercontent.com/13007381/194187165-46f8b366-4eea-4551-a2ec-32bfa42de741.mov

**After**

https://user-images.githubusercontent.com/13007381/194187122-1b98b06a-275f-4a5e-9695-a1e610fb73b5.mov

### TESTING INSTRUCTIONS

- Enable feature flag `DRILL_TO_DETAIL`
- Test the situations shown in the before/after videos above and make sure they behave as expected
- Make sure the modals open with the indicated filters when clicking "Drill to detail" and "Drill to detail by" options

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [x] Required feature flags: `DRILL_TO_DETAIL`
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/components/Chart/ChartContextMenu.tsx
Comment (zhaoyongjie) [inline superset-frontend/src/components/Chart/ChartContextMenu.tsx:44]: I suggest adding a flag on the chart's metadata instead of using a property, while the flag in the metadata refers to [this](https://github.com/apache/superset/blob/875e9f8a04464b0e79cfbda21d1e1ac7895970a5/superset-frontend/packages/superset-ui-core/src/chart/models/ChartMetadata.ts#L31).

**Review by zhaoyongjie — COMMENTED** [review]
left a quick suggestion

File: superset-frontend/src/components/Chart/ChartContextMenu.tsx
Comment (codyml) [reply superset-frontend/src/components/Chart/ChartContextMenu.tsx]: Done, thanks!

**Comment by michael-s-molina** [comment]
Thanks for the awesome PR description @codyml. I have a suggestion while looking at the screenshots. Can we standardize the way we display messages? 

For example:

193364518-6c079b56-aa36-4b57-968b-10533c4c8f45

I like disabling the option and showing an icon with a tooltip. Can we do the same for the "Drill to detail by" option? As you can see in the screenshot below, the message is displayed as a sub-item and the text is greyed out which is hard to read. It also forces the user to open the sub-menu before realizing the operation is not supported. I think using the icon plus tooltip also helps when the user already knows the possible causes and just wants to see if the option is enabled or not.
 
193364505-c38a9d43-565a-4088-b4f0-5f75094cbd23

**Comment by codyml** [comment]
> Thanks for the awesome PR description @codyml. I have a suggestion while looking at the screenshots. Can we standardize the way we display messages?
> 
> For example:
> 
> 193364518-6c079b56-aa36-4b57-968b-10533c4c8f45
> 
> I like disabling the option and showing an icon with a tooltip. Can we do the same for the "Drill to detail by" option? As you can see in the screenshot below, the message is displayed as a sub-item and the text is greyed out which is hard to read. It also forces the user to open the sub-menu before realizing the operation is not supported. I think using the icon plus tooltip also helps when the user already knows the possible causes and just wants to see if the option is enabled or not.
> 
> 193364505-c38a9d43-565a-4088-b4f0-5f75094cbd23

@michael-s-molina that works for me!  @kasiazjc any thoughts?

File: superset-frontend/plugins/plugin-chart-table/src/index.ts
Comment (codyml) [inline superset-frontend/plugins/plugin-chart-table/src/index.ts:54]: Have this check against an enum instead of a plain string.

File: superset-frontend/packages/superset-ui-core/src/chart/models/ChartMetadata.ts
Comment (geido) [inline superset-frontend/packages/superset-ui-core/src/chart/models/ChartMetadata.ts:20]: Let's use an absolute path to this. I believe we use absolute paths for whatever needs to go back of more than on directory.

File: superset-frontend/src/components/Chart/ChartRenderer.jsx
Comment (geido) [inline superset-frontend/src/components/Chart/ChartRenderer.jsx:86]: Do we have a better typescript definition for this to avoid using plain strings?

**Review by geido — COMMENTED** [review]
2 non-blocking nits

**Comment by geido** [comment]
Two nits from manual testing:

Probably not related to this PR but it looks weird that in the Modal there are two loading icons for both the table content and the Metadatabar cc @kasiazjc 

https://user-images.githubusercontent.com/60598000/194037386-8b882c9e-7fa6-4a75-acd2-82e021501398.mp4

It is possible to rightclick on a chart when the Modal is open. We should probably disable this possibility as the click isn't behaving correctly anyway

https://user-images.githubusercontent.com/60598000/194038602-96e2ef04-2f63-46bb-85f6-5d844d1abd6d.mp4

I thought we were supporting the Big number with trendline viz type but it shows that the Drill to detail by is disabled

**Comment by codyml** [comment]
> > @michael-s-molina that works for me! @kasiazjc any thoughts?
> 
> Awesome work, thanks @codyml ❤️
> 
> I think @michael-s-molina 's suggestion makes sense! let's go for it. Thanks for feedback Michael 🙏

@michael-s-molina @kasiazjc Done!  Updated videos in PR description.

File: superset-frontend/packages/superset-ui-core/src/chart/models/ChartMetadata.ts
Comment (codyml) [reply superset-frontend/packages/superset-ui-core/src/chart/models/ChartMetadata.ts]: Changed to `QueryFormData` for consistency and imported from `@superset-ui/core` – does that sound good?

File: superset-frontend/src/components/Chart/ChartRenderer.jsx
Comment (codyml) [reply superset-frontend/src/components/Chart/ChartRenderer.jsx]: Neither this component or the one that sets the prop is written in TS, but I was able to add an external module with an enum definition that I import and use.

**Comment by codyml** [comment]
> Two nits and a question from manual testing:
> 
> * Probably not related to this PR but it looks weird that in the Modal there are two loading icons for both the table content and the Metadatabar cc @kasiazjc
> 
>  Video.Game.Sales.mp4 

Agreed, will look into separately.

> * It is possible to rightclick on a chart when the Modal is open. We should probably disable this possibility as the click isn't behaving correctly anyway
> 
>  Video.Game.Sales.1.mp4 

Fixed.

> * I thought we were supporting the Big number with trendline viz type but it shows that the Drill to detail by is disabled, is that right?

Oops, turns out I forgot to add the `Behaviors.DRILL_TO_DETAIL` behavior to Big Number, Big Number w/ Trendline, Graph and Treemap v2, so they were all giving that error.  Should be fixed now.

**Review by michael-s-molina — COMMENTED** [review]
Thanks for the PR @codyml and for adding relevant comments to the code. It really helps with readability. I left some comments.

File: superset-frontend/src/components/Chart/ChartRenderer.jsx
Comment (codyml) [reply superset-frontend/src/components/Chart/ChartRenderer.jsx:300]: I think it's the same thing that's going on with World Map, where the React wrapper re-renders its contents whenever there's a prop change, so if we pass `inContextMenu` it causes the visualization to flash when you right-click.  You can see it happen if you add `Behavior.DRILL_TO_DETAIL` to one of the legacy charts (e.g. NVD3 Bar), or if you remove the behavior from World Map the flashing should stop.

File: superset-frontend/src/components/Chart/DrillDetail/DrillDetailMenuItems.test.tsx
Comment (codyml) [reply superset-frontend/src/components/Chart/DrillDetail/DrillDetailMenuItems.test.tsx:30]: For this file, I think that would result in a lot of duplicated code because I have a lot of `expect` calls in the helper functions: for example, 6 tests call `expectDrillToDetailEnabled`, which calls `expect` 5 times, so moving the `expect` calls to the top level would be 30 lines of code just for that check.  Plus, I made the find/expect logic pretty intertwined and it's not super straightforward to untangle it.  That being said, I mostly just don't really understand the reasoning behind the default settings on this lint rule – what's the argument for having `expect` calls only at the top level of each test?  It seems to me like having a rule against factoring out shared test logic encourages writing less thorough tests and makes tests more brittle and less readable, but if I'm missing something here I can go ahead and rewrite these to have `expect` at the top level.

File: superset-frontend/packages/superset-ui-core/src/chart/models/ChartMetadata.ts
Comment (codyml) [reply superset-frontend/packages/superset-ui-core/src/chart/models/ChartMetadata.ts]: Oops, fixed here and other places.

File: superset-frontend/src/components/Chart/ChartContextMenu.tsx
Comment (codyml) [reply superset-frontend/src/components/Chart/ChartContextMenu.tsx]: Done!

File: superset-frontend/src/components/Chart/ChartRenderer.jsx
Comment (codyml) [reply superset-frontend/src/components/Chart/ChartRenderer.jsx]: I think I did what you were suggestion, but let me know if that's not what you meant.

File: superset-frontend/src/components/Chart/DrillDetail/DrillDetailMenuItems.tsx
Comment (codyml) [reply superset-frontend/src/components/Chart/DrillDetail/DrillDetailMenuItems.tsx]: Updated!

File: superset-frontend/src/components/Chart/DrillDetail/DrillDetailMenuItems.tsx
Comment (zhaoyongjie) [inline superset-frontend/src/components/Chart/DrillDetail/DrillDetailMenuItems.tsx:134]: it's a `mean` component, should we rename it to `menu` or `drillToDetailMenu`?

File: superset-frontend/src/components/Chart/DrillDetail/DrillDetailMenuItems.tsx
Comment (zhaoyongjie) [inline superset-frontend/src/components/Chart/DrillDetail/DrillDetailMenuItems.tsx:153]: this one is the `menuItem` component so should rename to `menuItem` or `drillDetailMenuItem`.

File: superset-frontend/src/components/Chart/DrillDetail/DrillDetailMenuItems.tsx
Comment (zhaoyongjie) [inline superset-frontend/src/components/Chart/DrillDetail/DrillDetailMenuItems.tsx:173]: This is the third level if clause, should optimize readability.

**Review by zhaoyongjie — COMMENTED** [review]
Thanks for all of the new functions. I leave some comments. Good to hear your thought.

File: superset-frontend/src/types/ChartSource.ts
Comment (zhaoyongjie) [inline superset-frontend/src/types/ChartSource.ts:23]: one nite, the member of the enum is a constant so should use uppercase naming.

File: superset-frontend/src/components/Chart/ChartRenderer.jsx
Comment (michael-s-molina) [reply superset-frontend/src/components/Chart/ChartRenderer.jsx]: Oh got it! Thank you for the explanation 🙂

File: superset-frontend/src/components/Chart/DrillDetail/DrillDetailMenuItems.test.tsx
Comment (michael-s-molina) [reply superset-frontend/src/components/Chart/DrillDetail/DrillDetailMenuItems.test.tsx:30]: > That being said, I mostly just don't really understand the reasoning behind the default settings on this lint rule – what's the argument for having expect calls only at the top level of each test? It seems to me like having a rule against factoring out shared test logic encourages writing less thorough tests and makes tests more brittle and less readable, but if I'm missing something here I can go ahead and rewrite these to have expect at the top level.

The main reason for this lint rule is to prevent tests without assertions. Unfortunately, the rule is not smart enough to check helper functions. Another point is that unit tests generally don't share much logic so the default value makes sense for them. When we write integration tests like you're doing here, then shared logic is more common. In that case, it makes sense to disable the rule as you did if the code looks more readable 👍🏼

File: superset-frontend/packages/superset-ui-core/src/chart/models/ChartMetadata.ts
Comment (michael-s-molina) [reply superset-frontend/packages/superset-ui-core/src/chart/models/ChartMetadata.ts:53]: Automatically detecting if we have aggregations is way better. Nice suggestion @zhaoyongjie! 

Considering that `metrics` is defined as `metrics?: QueryFormMetric[];` you could use Lodash's `isEmpty` to improve readability. If we may have instances that disrespect Typescript, we could also use `ensureIsArray`.
```
const { metrics } = extractQueryFields(formData);
if (isEmpty(metrics)) {
  // logic for the noAggregations query.
}
```

File: superset-frontend/src/components/Chart/ChartContextMenu.tsx
Comment (michael-s-molina) [reply superset-frontend/src/components/Chart/ChartContextMenu.tsx:121]: I think we should keep `showDrillToDetail` in the dependencies array because the component does not know if `showDrillToDetail` will change. If in the future we make `isFeatureEnabled(FeatureFlag.DRILL_TO_DETAIL)` to dynamically return a value, the code will be ready for it. It will also trigger a warning if we remove the dependency.

File: superset-frontend/src/components/Chart/DrillDetail/DrillDetailMenuItems.tsx
Comment (michael-s-molina) [reply superset-frontend/src/components/Chart/DrillDetail/DrillDetailMenuItems.tsx:134]: Good point. I like the shorter version because we're already inside `DrillDetailMenuItems`. Unless there's a naming conflict 😉

File: superset-frontend/src/components/Chart/DrillDetail/DrillDetailMenuItems.tsx
Comment (michael-s-molina) [reply superset-frontend/src/components/Chart/DrillDetail/DrillDetailMenuItems.tsx:153]: Same as above.

File: superset-frontend/src/types/ChartSource.ts
Comment (michael-s-molina) [reply superset-frontend/src/types/ChartSource.ts:23]: Actually, Typescript recommends PascalCase for [enums](https://www.typescriptlang.org/docs/handbook/enums.html). You can check Ville's comment [here](https://github.com/apache/superset/pull/21438#discussion_r970821861).

**Comment by michael-s-molina** [comment]
@codyml Thanks for addressing all suggestions. I think after applying @zhaoyongjie suggestions we'll be in a good shape to involve QA as the final step before merging it. Thanks for all the hard work!

File: superset-frontend/plugins/plugin-chart-echarts/src/BigNumber/BigNumberViz.tsx
Comment (codyml) [reply superset-frontend/plugins/plugin-chart-echarts/src/BigNumber/BigNumberViz.tsx:29]: I think yes, because `DrillDetailTableControls` is currently only set up to correctly display binary filters (e.g. column = value), and if you change the type to `QueryObjectFilterClause` there you'll get typing errors that you can't resolve without adding new logic for displaying the other types.  Set filters and unary filters are definitely possible to add display support for in the future but in its current state it only works with binary filters.

File: superset-frontend/plugins/plugin-chart-table/src/types.ts
Comment (codyml) [reply superset-frontend/plugins/plugin-chart-table/src/types.ts]: Done!  Caught a few others too.

File: superset-frontend/src/components/Chart/ChartContextMenu.tsx
Comment (codyml) [reply superset-frontend/src/components/Chart/ChartContextMenu.tsx:76]: I don't think it's redundant?  I use that value in two places.

**Review by geido — APPROVED** [review]
Code LGTM. It would be great to have QA approval before merging

File: superset-frontend/src/components/Chart/ChartContextMenu.tsx
Comment (codyml) [reply superset-frontend/src/components/Chart/ChartContextMenu.tsx:105]: Agreed that this would work for the D2D case and is definitely cleaner, but I was trying to make it easy to add additional menu items to the contextual menu in the future (e.g. other drilling operations).  The way I have it now, you'd just add a new element for the menu items to `menuItems` and you'd add a new expression to the array that gets reduced into `itemsCount` and it should still work.  Do you think that's an over-optimization at this point though?

File: superset-frontend/src/components/Chart/ChartContextMenu.tsx
Comment (codyml) [reply superset-frontend/src/components/Chart/ChartContextMenu.tsx:121]: I'm with Michael on this one.

File: superset-frontend/src/components/Chart/ChartContextMenu.tsx
Comment (codyml) [reply superset-frontend/src/components/Chart/ChartContextMenu.tsx:136]: Is this necessary?  I feel like I've always seen `if (array.length) {...` as an example of concise/expressive JS, and there are way more instances of `if (array.length) {...` than `if (array.length > 0) {...` in the codebase.

File: superset-frontend/src/components/Chart/DrillDetail/DrillDetailMenuItems.tsx
Comment (codyml) [reply superset-frontend/src/components/Chart/DrillDetail/DrillDetailMenuItems.tsx:133]: `drillToDetail` is actually a menu item, so I could rename it `drillToDetailMenuItem` if you think that's clearer?  I don't think I should call it `menuItem` because `drillToDetailBy` is also a menu item.

File: superset-frontend/src/components/Chart/DrillDetail/DrillDetailMenuItems.tsx
Comment (codyml) [reply superset-frontend/src/components/Chart/DrillDetail/DrillDetailMenuItems.tsx]: Ok, renamed `drillToDetail` and `drillToDetailBy` to `drillToDetailMenuItem` and `drillToDetailByMenuItem`.

File: superset-frontend/src/types/ChartSource.ts
Comment (codyml) [reply superset-frontend/src/types/ChartSource.ts]: Updated to PascalCase.

File: superset-frontend/src/components/Chart/DrillDetail/DrillDetailMenuItems.tsx
Comment (codyml) [reply superset-frontend/src/components/Chart/DrillDetail/DrillDetailMenuItems.tsx]: Moved all conditionals to the top level – does that look better?

File: superset-frontend/src/components/Chart/ChartContextMenu.tsx
Comment (codyml) [reply superset-frontend/src/components/Chart/ChartContextMenu.tsx:121]: Even so, I think it's a good policy to keep things that React doesn't know won't change (e.g. everything that the lint rule picks up) in the dependency array.  I think the very small performance ding is worth it for component future-proofing (e.g. in case the definition of `isFeatureEnabled` changes) and keeping consistent with the lint rule.

File: superset-frontend/src/components/Chart/ChartContextMenu.tsx
Comment (zhaoyongjie) [reply superset-frontend/src/components/Chart/ChartContextMenu.tsx:136]: This is a convention in any language. of course, you can follow your logic. But what I want to say is that

```
vector.size > 0 // if the array is empty
vector.indexOf(item) > -1 // if a item in a Array
```

**Review by michael-s-molina — APPROVED** [review]
Code LGTM. Thanks for addressing all comments @codyml. Can we get @jinghua-qa approval before merging it?

**Pull request merged**
